codexdht/private/eth/p2p/discoveryv5/protocol.nim

```diff
@@ -537,7 +537,11 @@ proc ping*(d: Protocol, toNode: Node):
       discovery_message_requests_outgoing.inc(labelValues = ["invalid_response"])
       return err("Invalid response to ping message")
   else:
-    d.replaceNode(toNode)
+    # A ping (or the pong) was lost, what should we do? Previous implementation called 
+    # d.replaceNode(toNode) immediately, which removed the node. This is too aggressive,
+    # especially if we have a temporary network outage. Although bootstrap nodes are protected
+    # from being removed, everything else would slowly be removed.
+    d.replaceNode(toNode, removeIfNoReplacement = false)
```

Comment:
:yellow_circle: Hm... wouldn't it be better to perhaps make the removal more tolerant by, say, allowing a threshold level of responses to be missed for a peer? This way you may keep dead nodes around indefinitely?

Reply:
Yeah, I think we have to store more info and introduce some threshold. I'll do that before merging this.

Reply:
now storing more info (an exponential moving average of req/resp success), and having a drop threshold in case there is no replacement readily available.

Reply:
Note that I have also exposed a few other constants. Currently just exposing them without changing behavior, but I think we will want to eventually change these after careful evaluation.